File: api/schemas/bids.py
"""Pydantic schemas for bid API requests and responses."""

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class ConfigSummary(BaseModel):
    """Small config descriptor used by UI dropdowns."""

    id: str
    name: str
    path: str


class GenerateBidRequest(BaseModel):
    """Inputs required to build a bid preview or PDF package."""

    model_config = ConfigDict(populate_by_name=True)

    project_id: str = Field(..., examples=["shalom_prayer_center"])
    trade: str = Field("hvac", examples=["hvac", "plumbing", "electrical"])
    template_name: str | None = Field(None, examples=["bid_proposal"])
    package_name: Literal["all", "client", "internal"] = "all"
    region: str | None = None
    run_validation: bool = Field(True, alias="validate")


class LineItemResponse(BaseModel):
    cost_code: str
    description: str
    quantity: float
    unit: str
    unit_cost_material: float
    unit_cost_labor: float
    total_material: float
    total_labor: float
    total_phase: float
    labor_hours: float
    labor_factor: str
    sort_order: int


class BidTotalsResponse(BaseModel):
    total_material: float
    total_labor: float
    total_direct_cost: float
    contingency: float
    overhead_profit: float
    total_bid_amount: float
    total_labor_hours: float
    contingency_pct: float
    overhead_profit_pct: float


class ValidationIssueResponse(BaseModel):
    field: str
    message: str


class BidPreviewResponse(BaseModel):
    project_name: str
    bidder_name: str
    trade_name: str
    region: str
    status: str
    totals: BidTotalsResponse
    line_items: list[LineItemResponse]
    exclusions: list[str]
    validation: list[ValidationIssueResponse]


class GeneratedFileResponse(BaseModel):
    filename: str
    path: str
    url: str


class GenerateBidResponse(BaseModel):
    preview: BidPreviewResponse
    generated_files: list[GeneratedFileResponse]


class ProjectPricingResponse(BaseModel):
    """Project listing with real pricing data."""

    id: str
    name: str
    total_bid: float = 0.0
    total_material: float = 0.0
    total_labor: float = 0.0
    total_labor_hours: float = 0.0
    trade: str = "hvac"
    version_count: int = 0
    area_sf: float | None = None
    status: str = "estimating"
    archived: bool = False
    city: str | None = None
    state: str | None = None


class CreateProjectRequest(BaseModel):
    """Request to create a new project."""

    name: str
    city: str | None = None
    state: str | None = None
    trade: str = "hvac"
    total_area_sf: float | None = None
    construction_type: str | None = None


class BulkArchiveRequest(BaseModel):
    """Request to archive or unarchive multiple projects at once."""

    ids: list[str] = Field(..., examples=[["shalom_prayer_center", "office_building_a"]])
    archived: bool = Field(..., examples=[True])

File: frontend/src/types/bids.ts
/* ------------------------------------------------------------------ */
/*  Bid-related TypeScript types matching backend Pydantic schemas    */
/* ------------------------------------------------------------------ */

export interface GenerateBidRequest {
  project_id: string
  trade: string
  template_name?: string | null
  package_name?: 'all' | 'client' | 'internal'
  region?: string | null
  validate?: boolean
}

export interface LineItemResponse {
  cost_code: string
  description: string
  quantity: number
  unit: string
  unit_cost_material: number
  unit_cost_labor: number
  total_material: number
  total_labor: number
  total_phase: number
  labor_hours: number
  labor_factor: string
  sort_order: number
}

export interface BidTotalsResponse {
  total_material: number
  total_labor: number
  total_direct_cost: number
  contingency: number
  overhead_profit: number
  total_bid_amount: number
  total_labor_hours: number
  contingency_pct: number
  overhead_profit_pct: number
}

export interface ValidationIssueResponse {
  field: string
  message: string
}

export interface BidPreviewResponse {
  project_name: string
  bidder_name: string
  trade_name: string
  region: string
  status: string
  totals: BidTotalsResponse
  line_items: LineItemResponse[]
  exclusions: string[]
  validation: ValidationIssueResponse[]
}

export interface GeneratedFileResponse {
  filename: string
  path: string
  url: string
}

export interface GenerateBidResponse {
  preview: BidPreviewResponse
  generated_files: GeneratedFileResponse[]
}

/* ------------------------------------------------------------------ */
/*  Version snapshot types                                             */
/* ------------------------------------------------------------------ */

export interface CreateVersionRequest {
  trigger_source: 'user_edit' | 'ai_refine' | 'vision_import' | 'bulk_update'
  commit_message?: string | null
}

export interface TotalsDiff {
  from: number
  to: number
  delta: number
  delta_pct: number
}

export interface LineItemChange {
  cost_code: string
  description: string
  field: string
  from: unknown
  to: unknown
}

export interface VersionDiff {
  diff_type: string
  from_version: string | null
  to_version: string
  totals_changed: Record<string, TotalsDiff>
  line_items_changed: LineItemChange[]
  line_items_added: Record<string, unknown>[]
  line_items_removed: Record<string, unknown>[]
  summary: string
}

export interface SnapshotData {
  project_name: string
  trade_name: string
  totals: Record<string, unknown>
  line_items: Record<string, unknown>[]
  exclusions: string[]
}

export interface VersionMetadata {
  version_id: string
  timestamp: string
  commit_message: string
  trigger_source: string
  snapshot_summary?: string | null
}

export interface VersionListResponse {
  project_id: string
  trade: string
  versions: VersionMetadata[]
}

export interface VersionDetailResponse {
  version_id: string
  timestamp: string
  commit_message: string
  trigger_source: string
  snapshot_data: SnapshotData
  diff_from_previous: VersionDiff | null
}

export interface VersionDiffResponse {
  diff: VersionDiff | null
}

export interface CreateVersionResponse {
  version_id: string
  commit_message: string
  timestamp: string
  diff: VersionDiff | null
}

export interface RestoreVersionResponse {
  version_id: string
  project_name: string
  trade_name: string
  totals: Record<string, unknown>
  line_items: Record<string, unknown>[]
  exclusions: string[]
}

/* ------------------------------------------------------------------ */
/*  Trade / config summaries                                           */
/* ------------------------------------------------------------------ */

export interface ConfigSummary {
  id: string
  name: string
  path: string
}
